drivers/modem/modem_cellular.c

```diff
@@ -0,0 +1,1226 @@
+
+	atomic_set(&data->suspend_atomic, 0);
+	k_sem_init(&data->suspended_sem, 0, 1);
+
+	{
+		const struct modem_backend_uart_config uart_backend_config = {
+			.uart = config->uart,
+			.receive_buf = data->uart_backend_receive_buf,
+			.receive_buf_size = ARRAY_SIZE(data->uart_backend_receive_buf),
+			.transmit_buf = data->uart_backend_transmit_buf,
+			.transmit_buf_size = ARRAY_SIZE(data->uart_backend_transmit_buf),
+		};
+
+		data->uart_pipe = modem_backend_uart_init(&data->uart_backend,
+							  &uart_backend_config);
+	}
+
+	{
+		const struct modem_cmux_config cmux_config = {
+			.callback = modem_cellular_cmux_handler,
+			.user_data = data,
+			.receive_buf = data->cmux_receive_buf,
+			.receive_buf_size = ARRAY_SIZE(data->cmux_receive_buf),
+			.transmit_buf = data->cmux_transmit_buf,
+			.transmit_buf_size = ARRAY_SIZE(data->cmux_transmit_buf),
+		};
+
+		modem_cmux_init(&data->cmux, &cmux_config);
+	}
+
+	{
+		const struct modem_cmux_dlci_config dlci1_config = {
+			.dlci_address = 1,
+			.receive_buf = data->dlci1_receive_buf,
+			.receive_buf_size = ARRAY_SIZE(data->dlci1_receive_buf),
+		};
+
+		data->dlci1_pipe = modem_cmux_dlci_init(&data->cmux, &data->dlci1,
+							&dlci1_config);
+	}
+
+	{
+		const struct modem_cmux_dlci_config dlci2_config = {
+			.dlci_address = 2,
+			.receive_buf = data->dlci2_receive_buf,
+			.receive_buf_size = ARRAY_SIZE(data->dlci2_receive_buf),
+		};
+
+		data->dlci2_pipe = modem_cmux_dlci_init(&data->cmux, &data->dlci2,
+							&dlci2_config);
+	}
+
+	{
+		const struct modem_chat_config chat_config = {
+			.user_data = data,
+			.receive_buf = data->chat_receive_buf,
+			.receive_buf_size = ARRAY_SIZE(data->chat_receive_buf),
+			.delimiter = data->chat_delimiter,
+			.delimiter_size = ARRAY_SIZE(data->chat_delimiter),
+			.filter = data->chat_filter,
+			.filter_size = ARRAY_SIZE(data->chat_filter),
+			.argv = data->chat_argv,
+			.argv_size = ARRAY_SIZE(data->chat_argv),
+			.unsol_matches = NULL,
+			.unsol_matches_size = 0,
+			.process_timeout = K_MSEC(2),
+		};
+
+		modem_chat_init(&data->chat, &chat_config);
+	}
+
+#ifndef CONFIG_PM_DEVICE
+	modem_cellular_delegate_event(data, MODEM_CELLULAR_EVENT_RESUME);
+#else
+	pm_device_init_suspended(dev);
+#endif /* CONFIG_PM_DEVICE */
+
+	return 0;
+}
+
+#define MODEM_CELLULAR_INST_NAME(name, inst) \
+	_CONCAT(_CONCAT(_CONCAT(name, _), DT_DRV_COMPAT), inst)
+
+#define MODEM_CELLULAR_DEVICE(inst)								\
+	MODEM_PPP_DEFINE(MODEM_CELLULAR_INST_NAME(ppp, inst), NULL, 98, 1500, 64, 32);		\
+												\
+	static struct modem_cellular_data MODEM_CELLULAR_INST_NAME(data, inst) = {		\
+		.chat_delimiter = {'\r'},							\
+		.chat_filter = {'\n'},								\
+		.ppp = &MODEM_CELLULAR_INST_NAME(ppp, inst),					\
+	};											\
+												\
+	static struct modem_cellular_config MODEM_CELLULAR_INST_NAME(config, inst) = {		\
+		.uart = DEVICE_DT_GET(DT_INST_BUS(inst)),					\
+		.power_gpio = GPIO_DT_SPEC_INST_GET_OR(inst, mdm_power_gpios, {}),		\
```

Comment:
Add case for modems that don't have `power_gpios`. For example, HL7800 only has `reset_gpios`.